packages/style-engine/class-wp-style-engine-processor.php

```diff
@@ -65,29 +73,40 @@ public function add_rules( $css_rules ) {
 			}
 
 			foreach ( $css_rules as $rule ) {
-				$selector = $rule->get_selector();
-				$at_rule  = $rule->get_at_rule();
+				$rules_group = $rule->get_rules_group();
 
-				/**
-				 * If there is an at_rule and it already exists in the css_rules array,
-				 * add the rule to it.
-				 * Otherwise, create a new entry for the at_rule
+				/*
+				 * Merge existing rule and rules groups objects or create new ones.
+				 * Rules groups and rules are stored in separate arrays to allow for
+				 * unique processing.
 				 */
-				if ( ! empty( $at_rule ) ) {
-					if ( isset( $this->css_rules[ "$at_rule $selector" ] ) ) {
-						$this->css_rules[ "$at_rule $selector" ]->add_declarations( $rule->get_declarations() );
-						continue;
+				if ( $rule instanceof WP_Style_Engine_CSS_Rules_Group ) {
```

Comment:
This is the major change of this PR - processing rules groups and rules before CSS compilation mainly to:
- allow merging of nested rules, and therefore grouping nested rules under the same, single rules group selector/at-rule,
- allow for separate processing of groups, e.g., printing them at the bottom